### DISCUSSION TITLE
Issue with Identity Provider Token Exchange: "Client not allowed to exchange"

### DISCUSSION THREAD

**Discussion by BaselSalman** [keycloak/keycloak#23993] [Q&A]
I'm working on a feature where I need to exchange social tokens (e.g., from Facebook or Google) for Keycloak access tokens. I've configured the external social providers in Keycloak, and the process generally works as expected. However, when attempting to exchange a token from a user that's already linked with the respective identity provider, I encounter a 403 status code with the error message "Client not allowed to exchange".

Here's the relevant request body (as per the provided screenshot):

![image](https://github.com/keycloak/keycloak/assets/107203501/13e9c36e-4b09-43ad-810b-c276b0de8e25)

I've ensured that my configuration aligns with Keycloak's documentation, and I'm at a bit of a loss regarding this error.

Has anyone faced a similar issue? Any insights or solutions would be greatly appreciated.

**Comment by dsofachats** [comment]
Hello there.
I'm encountering the exact same issue. @BaselSalman were you able to solve this issue?

**Reply by BaselSalman** [reply]
Hey, you must enable the token exchange preview feature as written in docs:
**https://www.keycloak.org/server/features**

You need to bash inside your docker container -if using docker- and type the command: **kc.sh build --features="preview"**. If you're using a custom dockerfile image, you need to add this in your dockerfile: 
**RUN /opt/keycloak/bin/kc.sh build --features="preview"**

**Reply by Jeff-Tian** [reply]
No, when the feature is not enabled, the error is "feature not enabled". The error can only be "Client not allowed to exchange" when this feature is enabled. 

<img width="3728" height="1300" alt="image" src="https://github.com/user-attachments/assets/802dcc39-072f-4c79-90bd-8550e5061378" />

To double-check if it's enabled, you can go to the admin console and confirm it's listed:
<img width="2578" height="1496" alt="image" src="https://github.com/user-attachments/assets/0f54512f-a1de-4ff2-b8e4-975a03ae690e" />

**Comment by TH3AL3X** [comment]
Anyone here solved?

**Comment by Jeff-Tian** [comment]
> Anyone here solved?

I successfully enabled the token exchange feature and made it work in my case. I'd like to share the information below, hoping it can help resolve your problem as well.

### 1. First, make sure you have enabled the feature as pointed out here: https://github.com/keycloak/keycloak/discussions/23993#discussioncomment-12432595
### 2.  Second, make sure you have enabled "Fine-grained permissions" for the client that'll initiate the token exchange flow. If not enabled, it looks like: 
<img width="2150" height="1006" alt="image" src="https://github.com/user-attachments/assets/7427a823-d2ad-4abb-859e-ba7b0159b8dc" />
After being enabled, it looks like: 
<img width="2376" height="1426" alt="image" src="https://github.com/user-attachments/assets/a1ade4bc-94d8-479c-a5bb-e8b0c285afee" />, noticing the new tabs appeared with the "token exchange" item.

### 3. Then, add a policy to allow the token exchange flow. You need to go to the "realm-management" client to do that, not the client that'll initiate the flow.
<img width="2186" height="592" alt="image" src="https://github.com/user-attachments/assets/0ab66605-acf3-469c-8ee7-ecb4039412a4" />
Instead, you need to click the "token-exchange" link in the previous tab:
<img width="2226" height="1288" alt="image" src="https://github.com/user-attachments/assets/4dd23c5c-cb95-4e22-abbc-0191c831185e" />
Then you specify the "token-exchange" and then create a policy: 
<img width="1722" height="1476" alt="image" src="https://github.com/user-attachments/assets/8eca169a-92e2-4724-8b3a-4e0b8203bd45" />, and allow the token exchange initiated by the previous client:
<img width="1762" height="1352" alt="image" src="https://github.com/user-attachments/assets/082f6b05-954f-48f3-b99f-9dc7d3406f7b" />

### 4. Back to the previous "permission details" tab and apply the policy to the permission:
<img width="1794" height="1472" alt="image" src="https://github.com/user-attachments/assets/7a625f18-7c6f-4bc7-907f-af39e5b5d0f8" />

### 5. Now you are good to test the token exchange flow: 
<img width="2740" height="1582" alt="image" src="https://github.com/user-attachments/assets/35024bbb-1d7b-4fec-aaae-914ccc5853e9" />

**Reply by Lamotato** [reply]
Hi! Thank you so much for your help. I have followed your configuration instructions, but the issue still persists. I am using Keycloak version 23.0.7. Are there any additional configurations or steps I need to complete? Thank you very much!
![Uploading QQ20260429-011058.png…]()

**Reply by Jeff-Tian** [reply]
如果需要更多帮助，欢迎通过 https://www.zhihu.com/consult/people/1073548674713423872 咨询。
